--- mpss-modules-3.8.4/host/linux.c	2016-05-31 09:48:37.000000000 +0200
+++ mpss-modules-3.8.4jjk/host/linux.c	2018-05-17 15:28:56.139635516 +0200
@@ -337,7 +337,7 @@
 			NULL, "mic%d", mic_ctx->bd_info->bi_ctx.bi_id);
 	err = sysfs_create_group(&mic_ctx->bd_info->bi_sysfsdev->kobj, &bd_attr_group);
 	mic_ctx->sysfs_state = sysfs_get_dirent(mic_ctx->bd_info->bi_sysfsdev->kobj.sd,
-#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,35) && LINUX_VERSION_CODE < KERNEL_VERSION(3,14,0))
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,35) && LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0))
 				NULL,
 #endif
 				"state");
--- mpss-modules-3.8.4/host/linvnet.c	2016-05-31 09:48:37.000000000 +0200
+++ mpss-modules-3.8.4jjk/host/linvnet.c	2018-05-17 15:45:08.597678495 +0200
@@ -203,7 +203,7 @@
 	.ndo_set_multicast_list = micveth_multicast_list,
 #endif
 	.ndo_set_mac_address	= micveth_set_address,
-	.ndo_change_mtu		= micveth_change_mtu,
+	.ndo_change_mtu_rh74	= micveth_change_mtu,
 };
 #endif
 
--- mpss-modules-3.8.4/vnet/micveth_dma.c	2016-05-31 09:48:37.000000000 +0200
+++ mpss-modules-3.8.4jjk/vnet/micveth_dma.c	2018-05-17 15:46:17.078060653 +0200
@@ -910,7 +910,7 @@
 	.ndo_set_multicast_list = micvnet_multicast_list,
 #endif
 	.ndo_set_mac_address	= micvnet_set_address,
-	.ndo_change_mtu		= micvnet_change_mtu,
+	.ndo_change_mtu_rh74	= micvnet_change_mtu,
 };
 
 static void
